File: python/selene_isru/modules/energy_ledger.py
from __future__ import annotations

import sys
from typing import Any

from ..constants import c


def _balance(node_id: str, label: str, electrical: float, coupled: float, useful: float, rejected: float, accumulation: float) -> dict[str, Any]:
    raw = electrical + coupled - useful - rejected - accumulation
    scale = max(1, abs(electrical), abs(coupled), abs(useful), abs(rejected), abs(accumulation))
    roundoff_tolerance_w = 64 * sys.float_info.epsilon * scale
    return {
        "id": node_id, "label": label, "electricalInputW": electrical, "coupledInputW": coupled,
        "usefulOutputW": useful, "rejectedHeatW": rejected, "accumulationW": accumulation,
        "residualW": 0 if abs(raw) <= roundoff_tolerance_w else raw,
    }


def energy_ledger(
    params: dict[str, Any],
    grid_power_w: float,
    lines: list[dict[str, Any]],
    excavation_mech_power_w: float,
    electrolysis: dict[str, Any],
    cryo: dict[str, Any],
    sabatier: dict[str, Any] | None,
) -> dict[str, Any]:
    mass_flow = params["targetKgPerDay"] / 86400

    def power_for(source: str, target: str) -> float:
        line = next((item for item in lines if item["from"] == source and item["to"] == target), None)
        return 0 if line is None else line["jPerKg"] * mass_flow

    balances: list[dict[str, Any]] = []
    excavation_input = power_for("mine", "melt" if params["site"] == "equatorial" else "sublimation")
    excavation_useful = min(excavation_input, max(0, excavation_mech_power_w))
    balances.append(_balance("excavation-energy", "Excavation drive", excavation_input, 0, excavation_useful, excavation_input - excavation_useful, 0))
    if params["site"] == "equatorial":
        melt_input = power_for("melt", "electrolysis")
        balances.append(_balance("mre-melt-energy", "Regolith melt duty", melt_input, 0, 0, 0, melt_input))
        electrolysis_input = power_for("electrolysis", "product")
        chemical = min(electrolysis_input, max(0, electrolysis["chemicalPowerW"]))
        balances.append(_balance("mre-electrolysis-energy", "MRE voltage and reaction", electrolysis_input, 0, chemical, electrolysis_input - chemical, 0))
        auxiliaries = power_for("electrolysis", "parasitic")
        balances.append(_balance("mre-aux-energy", "MRE radiation and auxiliaries", auxiliaries, 0, 0, auxiliaries, 0))
    else:
        sublimation_input = power_for("sublimation", "product")
        balances.append(_balance("sublimation-energy", "Polar heating and sublimation", sublimation_input, 0, 0, 0, sublimation_input))
        distillation = power_for("sublimation", "parasitic")
        balances.append(_balance("polar-aux-energy", "Vapor handling and process allowance", distillation, 0, 0, distillation, 0))
        if sabatier is not None:
            electrolysis_input = power_for("electrolysis", "product")
            fraction = min(1, c("VthermoneutralWater") * params["etaFaradayEl"] / params["Vel"])
            chemical = electrolysis_input * fraction
            balances.append(_balance("water-electrolysis-energy", "Water electrolysis", electrolysis_input, 0, chemical, electrolysis_input - chemical, 0))
    balances.append(_balance("storage-conditioning-energy", "Product conditioning", cryo["totalConditioningPowerW"], 0, 0, 0, cryo["totalConditioningPowerW"]))
    balances.append(_balance("storage-cooling-energy", "Storage heat lift", cryo["cryocoolerPowerW"], cryo["qRemovedW"], 0, cryo["cryocoolerPowerW"] + cryo["qRemovedW"], 0))
    allocated = sum(item["electricalInputW"] for item in balances)
    raw_grid_residual = grid_power_w - allocated
    allocation_scale = max(1, abs(grid_power_w), abs(allocated))
    grid_residual = 0 if abs(raw_grid_residual) <= 64 * sys.float_info.epsilon * allocation_scale else raw_grid_residual
    maximum = max([abs(grid_residual), *(abs(item["residualW"]) for item in balances)])
    return {"balances": balances, "maxAbsResidualW": maximum, "gridAllocationResidualW": grid_residual}

File: packages/engine/src/modules/energyLedger.ts
import { PHYSICAL_CONSTANTS } from "../constants";
import type { EnergyProcessBalance, SimParams } from "../types";
import type { CryoOutput } from "./cryo";
import type { ElectrolysisOutput } from "./electrolysis";

interface EnergyLine {
  from: string;
  to: string;
  jPerKg: number;
}

interface SabatierEnergy {
  secWaterElectrolysis_JPerKg: number;
}

export interface EnergyLedger {
  balances: EnergyProcessBalance[];
  maxAbsResidualW: number;
  gridAllocationResidualW: number;
}

function balance(
  id: string,
  label: string,
  electricalInputW: number,
  coupledInputW: number,
  usefulOutputW: number,
  rejectedHeatW: number,
  accumulationW: number
): EnergyProcessBalance {
  const raw = electricalInputW + coupledInputW - usefulOutputW - rejectedHeatW - accumulationW;
  const scale = Math.max(
    1,
    Math.abs(electricalInputW),
    Math.abs(coupledInputW),
    Math.abs(usefulOutputW),
    Math.abs(rejectedHeatW),
    Math.abs(accumulationW)
  );
  const roundoffToleranceW = 64 * Number.EPSILON * scale;
  return {
    id,
    label,
    electricalInputW,
    coupledInputW,
    usefulOutputW,
    rejectedHeatW,
    accumulationW,
    residualW: Math.abs(raw) <= roundoffToleranceW ? 0 : raw
  };
}

export function energyLedger(
  params: SimParams,
  gridPowerW: number,
  lines: EnergyLine[],
  excavationMechPowerW: number,
  electrolysis: ElectrolysisOutput,
  cryo: CryoOutput,
  sabatier: SabatierEnergy | null
): EnergyLedger {
  const productMassFlowKgPerS = params.targetKgPerDay / 86_400;
  const powerFor = (from: string, to: string): number =>
    (lines.find((line) => line.from === from && line.to === to)?.jPerKg ?? 0) * productMassFlowKgPerS;
  const balances: EnergyProcessBalance[] = [];

  const excavationInputW = powerFor("mine", params.site === "equatorial" ? "melt" : "sublimation");
  const excavationUsefulW = Math.min(excavationInputW, Math.max(0, excavationMechPowerW));
  balances.push(balance("excavation-energy", "Excavation drive", excavationInputW, 0, excavationUsefulW, excavationInputW - excavationUsefulW, 0));

  if (params.site === "equatorial") {
    const meltInputW = powerFor("melt", "electrolysis");
    balances.push(balance("mre-melt-energy", "Regolith melt duty", meltInputW, 0, 0, 0, meltInputW));

    const electrolysisInputW = powerFor("electrolysis", "product");
    const chemicalW = Math.min(electrolysisInputW, Math.max(0, electrolysis.chemicalPowerW));
    balances.push(balance("mre-electrolysis-energy", "MRE voltage and reaction", electrolysisInputW, 0, chemicalW, electrolysisInputW - chemicalW, 0));

    const auxiliariesW = powerFor("electrolysis", "parasitic");
    balances.push(balance("mre-aux-energy", "MRE radiation and auxiliaries", auxiliariesW, 0, 0, auxiliariesW, 0));
  } else {
    const sublimationInputW = powerFor("sublimation", "product");
    balances.push(balance("sublimation-energy", "Polar heating and sublimation", sublimationInputW, 0, 0, 0, sublimationInputW));
    const distillationW = powerFor("sublimation", "parasitic");
    balances.push(balance("polar-aux-energy", "Vapor handling and process allowance", distillationW, 0, 0, distillationW, 0));

    if (sabatier !== null) {
      const waterElectrolysisW = powerFor("electrolysis", "product");
      const thermoneutralFraction = Math.min(
        1,
        PHYSICAL_CONSTANTS.VthermoneutralWater.value * params.etaFaradayEl / params.Vel
      );
      const chemicalW = waterElectrolysisW * thermoneutralFraction;
      balances.push(balance("water-electrolysis-energy", "Water electrolysis", waterElectrolysisW, 0, chemicalW, waterElectrolysisW - chemicalW, 0));
    }
  }

  balances.push(balance("storage-conditioning-energy", "Product conditioning", cryo.totalConditioningPowerW, 0, 0, 0, cryo.totalConditioningPowerW));
  balances.push(balance("storage-cooling-energy", "Storage heat lift", cryo.cryocoolerPowerW, cryo.qRemovedW, 0, cryo.cryocoolerPowerW + cryo.qRemovedW, 0));

  const allocatedElectricalW = balances.reduce((total, item) => total + item.electricalInputW, 0);
  const gridAllocationRaw = gridPowerW - allocatedElectricalW;
  const allocationScale = Math.max(1, Math.abs(gridPowerW), Math.abs(allocatedElectricalW));
  const gridAllocationResidualW =
    Math.abs(gridAllocationRaw) <= 64 * Number.EPSILON * allocationScale ? 0 : gridAllocationRaw;
  const maxAbsResidualW = balances.reduce((maximum, item) => Math.max(maximum, Math.abs(item.residualW)), Math.abs(gridAllocationResidualW));
  return { balances, maxAbsResidualW, gridAllocationResidualW };
}
